add Chai's pathval features to this separate module

lucasfcosta opened the pull request:
This PR holds the progress I've made on moving Chai's `pathval` features to this separate module.
I have done some changes to the code brought here in order to have 100% coverage and follow the linting rules defined.

This is almost done, but I'd still like to refactor some parts of the code, such as turning the `internalSetPathValue` function into a recursive function and maybe renaming some variables for them to make sense.

If you have any other refactoring suggestions please let me know.
I'll remove the `[WIP]` tag and do another post here when I finish this one.

**Important Detail:**
[Currently the build is failing for Node `0.8` because `type-detect` (which is a dependency of `pathval`) needs a Node version above `0.1.0`.](https://travis-ci.org/chaijs/pathval/jobs/141903679#L230) What are we going to do about it? According to [the `type-detect` package.json](https://github.com/chaijs/type-detect/blob/master/package.json#L83) it should be compatible with any Node.js version as soon as the version on the repo gets released.

Thanks for your attention 😄 

---------------------------------------------------------------------
EDIT: This one is done. Please read the comment below.

lucasfcosta commented:
Hey everyone, this one is done!
I thought it would be good to refactor `getPathValue` and make it a recursive method instead of using a `for` loop, but after quite some time thinking about it I thought it would be better not to mess with it because we've gotta track the state of the properties we need and this means we gotta have an index for then, so passing the same `parsed` object around wouldn't do much good and would only add unnecessary complexity to the code. If you guys really think this is necessary let me know, but I think what we've got for now is sufficient.

I would also like to ask you if I should add two methods called `.get` and `.set` that call `getPathValue` and `setPathValue` in order not to break backwards compatibility with the old API.

Please let me know if you find any problems with this

PS.: Please read the note above about the NodeJS version required for this package

meeber commented on README.md:
```diff
@@ -1,102 +1,208 @@
+
+The primary export of `pathval` is an object which has the following methods:
+
+// TODO
```
todo :D

lucasfcosta replied:
Oops, forgot that, srry, let me take another look at other things than code hahaha

meeber commented on index.js:
```diff
@@ -1,189 +1,317 @@
  *
  * ### Paths:
  *
  * * Can be as near infinitely deep and nested
```
"Can be as near infinitely deep" sounds off linguistically. Maybe just remove "as near"? I think it's implied that that memory limitations would prevent it from actually being infinitely deep.

lucasfcosta replied:
Yes, makes sense.
I also noticed that I forgot to update some other strings here I have changed before but reseted my branch.
I'll push changes in minutes.

lucasfcosta commented:
Now it's really done!
This is a really big one, so let me know if I missed any typos or anything.
Thanks for the help @meeber 😄

meeber commented on README.md:
```diff
@@ -1,102 +1,146 @@
+The primary export of `pathval` is an object which has the following methods:
 
-props.get(arr, '[0].foo'); // => "bar"
+// TODO
```
Still says TODO though ;)

lucasfcosta replied:
oops

meeber commented on index.js:
```diff
@@ -1,189 +1,317 @@
  *
  * @param {Object} parsed definition from `parsePath`.
  * @param {Object} object to search against
+ * @param {Number} object to search against
```
Description on 115 needs updating

meeber commented on index.js:
```diff
@@ -1,189 +1,317 @@
+  pathDepth = (typeof pathDepth === 'undefined' ? parsed.length : pathDepth);
 
-  for (var i = 0, l = parsed.length; i < l; i++) {
+  for (var i = 0, pathPart = pathDepth; i < pathPart; i++) {
```
What's the purpose of `pathPart`? Why can't `pathDepth` be used directly instead?

lucasfcosta replied:
You're right, we can use it since we're not even setting that value.

meeber commented on index.js:
```diff
@@ -1,189 +1,317 @@
+ */
+
+function defined(val) {
+  return !(!val && typeof val === 'undefined');
```
Hmm. Does this do the same thing?:

```js
return typeof val !== 'undefined';
```

lucasfcosta replied:
Apparently it does.
I have seen this at the first time I was reading the code and then I thought that there could be a good reason for this to be there, so I realized that the cost of performing `!val` (just accessing a single property and turning it into an inverse boolean) is lower than  executing typeof. So whenever `!val` returns true the `typeof` won't even be called.

IMO the difference between both is almost none, we can change this to what you suggested, what do you think @meeber?

meeber replied:
I prefer the simple form but if you prefer it this way that's cool, just add a small comment that it's for optimization.

keithamus replied:
Better than a comment - if it makes a difference, prove it with a benchmark that we can re-run (like the ones in type-detect). If it doesn't make a (significant) difference, then ditch it.

meeber commented on index.js:
```diff
@@ -1,189 +1,317 @@
+    // If it's the last part of the path, we set the 'x' value with the property name
+    if (defined(tempObj) && i === (pathDepth - 1)) {
+      propName = defined(part.p) ? part.p : part.i;
+      // Now we set the property with the name held by 'x' on object with the desired val
```
L181 and L184: 'x' isn't the name of the var anymore

lucasfcosta replied:
Thanks for spotting these!
Fixed.

meeber commented on index.js:
```diff
@@ -1,189 +1,317 @@
+      tempObj[propName] = val;
+    } else if (defined(tempObj) && defined(part.p) && tempObj[part.p]) {
+      tempObj = tempObj[part.p];
+    } else if (defined(tempObj) && defined(part.i) && tempObj[part.i]) {
```
I know this is something that existed previously, not something you added, but I'm just wondering what exactly the `defined(obj)` on lines 182, 186, and 188 are protecting against? I tried removing them and all the tests still pass.

lucasfcosta replied:
Yeah, you are right, since we already got this condition on the clause `tempObj[part.p]` we won't need `defined(tempObj)`. The only way it would be undefined was if the object passed to the function was `undefined`, and in that case it would fail anyway.
I'll remove it today and `amend` into this commit.
Thanks for noticing.

meeber commented on test/index.js:
```diff
@@ -0,0 +1,211 @@
+    assert(obj.hello.universe[1].properties === 'galaxy');
+  });
+
+  it('allows value to be REset in simple object', function () {
```
Is there significance of capitalization in `REset` or is that just a search/replace side-effect?

lucasfcosta replied:
Yes, it was already like that before this change so I thought it would be better to leave it like this.
It aims at emphasizing we are setting a value again (`RE`setting).

meeber commented:
@lucasfcosta Looks like you've been busy lately. Great work! I think this module has a lot of value being standalone; the string syntax of property access is pretty nifty :D 

In addition to the line comments above, I have a couple questions:

- I find the parameter ordering a bit surprising in many of the functions. In particular, I'd expect `obj` to be the first parameter instead of the last. It looks like `obj` did use to be first in `.set`. and `.get`. Was the change intentional?
- What's with `test.sh`?

lucasfcosta commented:
@meeber Yeah, I took those functions out of Chai core and they had this parameter order, so I thought it should've been kept the same, that's why [I had the doubt if I should or add `get` and `set` methods with the same signatures as before just to keep the API compatible with the previous version](https://github.com/chaijs/pathval/pull/11#issuecomment-231501186).

`test.sh` is a test file I was using in order to run tests on my IE VMs and make sure everything was working on IE too, I better remove that, well noticed.

I have also added a comment explaining why `.defined(val)` exists because I also thought it keeps the code cleaner, it's better to write `defined(val)` at a lot of places than having to use `typeof` every single time. What do you think about this?

Thanks for reviewing, it's always good to have another pair of eyes on such a big PR.

meeber commented:
@lucasfcosta Adding the `get` and `set` methods like before would be good. If this were a new project, I'd say all of the methods should accept `obj` as their first parameter, but it's probably not worth introducing such a breaking change after the fact.

keithamus commented:
Having all of the methods take the object first seems sensible, right now they're inconsistent and thats, IMO, the worst case scenario. 

Remember this (and chai) will be a breaking version so - while we should make the effort to ensure this doesn't break lots of plugins - we also have license to change things if it makes the libs better, which I think unequivocally it does here.

lucasfcosta commented:
Here it is. Thanks for your input guys 😄 
I changed all the methods to take the `object` first and then the `path`.
I also updated all the docs (including `README.md`) and the docStrings' examples

I also removed some unnecessary `defined` checks [as noticed by @meeber here](https://github.com/chaijs/pathval/pull/11#discussion_r70160880).

Since @keithamus said this will be a breaking version I didn't add the old `set` and `get` methods, the ones on this PR have the same behavior as the old ones even though they have different names.

I think this is good to go, please let me know if you find anything else that needs to be changed.

meeber commented:
@lucasfcosta Having all of the methods take `object` as their first parameter brings me far greater happiness than such a small thing ever should.

This looks just about done to me. I'm fine with no `set` and `get` methods given the available equivalents.

There does seem to be one outstanding item regarding @keithamus' note about benchmarking the `defined` helper function.

lucasfcosta commented:
@meeber Thanks for pointing it out, I haven't seen that comment!
I'll do it 😄

lucasfcosta commented:
Hello everyone, well I've used [this module](https://www.npmjs.com/package/benchmark) and I have created two versions of the `internalSetPathValue` and `setPathValue` to measure the performance of the `defined` method and `typeof obj !== 'undefined' and I have discovered that `defined` is a little bit faster than using `typeof`. This is not a much significant difference and one time the `typeof version was a little bit faster.

Here are the results I have collected:

```
# Result 1
Without defined() x 278,315 ops/sec ±11.08% (67 runs sampled)
With defined() x 280,066 ops/sec ±10.67% (72 runs sampled)

# Result 2
Without defined() x 280,427 ops/sec ±9.83% (70 runs sampled)
With defined() x 293,800 ops/sec ±7.80% (77 runs sampled)

# Result 3
Without defined() x 280,420 ops/sec ±8.44% (74 runs sampled)
With defined() x 278,014 ops/sec ±8.95% (75 runs sampled)

# Result 4
Without defined() x 231,060 ops/sec ±14.99% (60 runs sampled)
With defined() x 267,356 ops/sec ±9.32% (67 runs sampled)
```

I also ran this a couple more times and got even less significant results, these are the four first executions, which seem to summarize the situation pretty well.

So, as @keithamus suggested I'm going to drop this method.

Thank you guys for making me dig deeper in this. This reminds me of [the first of the Akin's Laws of
Spacecraft Design](http://spacecraft.ssl.umd.edu/akins_laws.html), which states that:

> 1. Engineering is done with numbers. Analysis without numbers is only an opinion.

Well, that was quite a lesson to learn today, have a nice sunday folks 😄

lucasfcosta commented:
Ping @keithamus @meeber 
Hey guys, I'm sorry to bother but I'll be going to spend my vacation abroad so I may not be able to commit between July 28th and Aug 12th, so if you guys need anything regarding this one (or any other thing) please let me know on this PR (or tag me in the issue you need me) and I'll do my best to get things going again as soon as I get back from the trip. 😄 

However, until July 28th, you can count on me 👍

meeber commented:
@lucasfcosta Vacation? I don't remember approving a vacation.

@keithamus Did you approve this vacation?

Regarding this PR: LGTM

lucasfcosta commented:
@meeber Oh I'm sorry Mr. Meeber, I talked with the Internet's Human Resources department and they said it was okay.
I hope I could send you some pizza straight from Italy but I doubt it will get to Virginia before it gets too cold.

keithamus commented:
Okay, sorry for holding this one up. LGTM. Let's merge it and get working on releasing 😄

keithamus commented:
Also, I most certainly did not approve of any vacation! Our vacation policy is very strict: you're only allowed a max of 365 days a year and all vacation will go unpaid.

CacheControl commented:
@lucasfcosta @keithamus 

Can you provide an update on any plans for publishing a new major version?  As this repo currently stands, it's very confusing to `npm install` and receive as a completely different api than what the github README shows (e.g. `pathvalue.set()` exists instead of `pathvalue. setPathValue()`

On that note, was there any particular reason to rename `.get()` and `.set()`?  If the functionality hasn't changed, it'd be nice to at least alias those methods to their new counterparts for providing backwards compatibility with the previous api.

keithamus commented:
Hey @CacheControl. You can check the website for the docs for the latest version of chai. We'll be releasing 4.0 pretty soon, and it'll be a major breaking version, which means api changes like what you've described will happen. We will be releasing comprehensive changelogs with the release of 4.0 so you know what has changed.
